frankenstein/consul_client.go

```diff
@@ -0,0 +1,168 @@
+package frankenstein
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	consul "github.com/hashicorp/consul/api"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// ConsulClient is a high-level client for Consul, that exposes operations
+// such as CAS and Watch which take callbacks.  It also deals with serialisation.
+type ConsulClient interface {
+	Get(key string, out interface{}) error
+	CAS(key string, out interface{}, f CASCallback) error
+	WatchPrefix(prefix string, out interface{}, done chan struct{}, f func(string, interface{}) bool)
+}
+
+// CASCallback is the type of the callback to CAS.  If err is nil, out must be non-nil.
+type CASCallback func(in interface{}) (out interface{}, retry bool, err error)
+
+// NewConsulClient returns a new ConsulClient
+func NewConsulClient(addr string) (ConsulClient, error) {
+	client, err := consul.NewClient(&consul.Config{
+		Address: addr,
+		Scheme:  "http",
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &consulClient{client.KV()}, nil
+}
+
+var (
+	queryOptions = &consul.QueryOptions{
+		RequireConsistent: true,
+	}
+	writeOptions = &consul.WriteOptions{}
+
+	// ErrNotFound is returned by ConsulClient.Get
+	ErrNotFound = fmt.Errorf("Not found")
+)
+
+type kv interface {
+	CAS(p *consul.KVPair, q *consul.WriteOptions) (bool, *consul.WriteMeta, error)
+	Get(key string, q *consul.QueryOptions) (*consul.KVPair, *consul.QueryMeta, error)
+	List(prefix string, q *consul.QueryOptions) (consul.KVPairs, *consul.QueryMeta, error)
+}
+
+type consulClient struct {
+	kv
+}
+
+// Get and deserialise a JSON value from consul.
+func (c *consulClient) Get(key string, out interface{}) error {
+	kvp, _, err := c.kv.Get(key, queryOptions)
+	if err != nil {
+		return err
+	}
+	if kvp == nil {
+		return ErrNotFound
+	}
+	if err := json.NewDecoder(bytes.NewReader(kvp.Value)).Decode(out); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CAS atomically modify a value in a callback.
+// If value doesn't exist you'll get nil as a argument to your callback.
+func (c *consulClient) CAS(key string, out interface{}, f CASCallback) error {
+	var (
+		index        = uint64(0)
+		retries      = 10
+		retry        = true
+		intermediate interface{}
+	)
+	for i := 0; i < retries; i++ {
+		kvp, _, err := c.kv.Get(key, queryOptions)
+		if err != nil {
+			log.Errorf("Error getting %s: %v", key, err)
+			continue
+		}
+		if kvp != nil {
+			if err := json.NewDecoder(bytes.NewReader(kvp.Value)).Decode(out); err != nil {
+				log.Errorf("Error deserialising %s: %v", key, err)
+				continue
+			}
+			index = kvp.ModifyIndex // if key doesn't exist, index will be 0
```

Comment:
nit: Any comments in Prometheus should be sentence-style (start with a capital letter, end with period).